widgets/src/lib/InferenceWidget/widgets/StructuredDataClassificationWidget/StructuredDataClassificationWidget.svelte

```diff
@@ -0,0 +1,204 @@
+	export let shouldUpdateUrl: WidgetProps["shouldUpdateUrl"];
+
+	const columns: string[] = Object.keys(
+		model?.widgetData[0]?.structuredData ?? {}
+	);
+
+	let computeTime = "";
+	let error: string = "";
+	let isLoading = false;
+	let modelLoading = {
+		isLoading: false,
+		estimatedTime: 0,
+	};
+	let output: Output | null = null;
+	let outputJson: string;
+	let table: string[][] = [columns];
+	let query = "How many stars does the transformers repository have?";
+
+	onMount(() => {
+		const [dataParam] = getSearchParams(["structuredData"]);
+		if (dataParam) {
+			table = convertDataToTable((parseJSON(dataParam) as TableData) ?? {});
+			getOutput();
+		} else {
+			const [demoTable] = getDemoInputs(model, ["structuredData"]);
+			table = convertDataToTable(demoTable as TableData);
+			if (query && table && callApiOnMount) {
+				getOutput();
+			}
+		}
+	});
+
+	function onChangeTable(updatedTable: string[][]) {
+		table = updatedTable;
+	}
+
+	async function getOutput(withModelLoading = false) {
+		if (table?.[0].length !== columns.length) {
+			error = `Data needs to have ${columns.length} columns`;
+			output = null;
+			outputJson = "";
+			return;
+		}
+
+		if (shouldUpdateUrl) {
+			updateUrl({
+				data: JSON.stringify(convertTableToData(table)),
+			});
+		}
+
+		const requestBody = {
+			inputs: {
+				data: convertTableToData(table),
+			},
+		};
+
+		isLoading = true;
+
+		const res = await getResponse(
+			apiUrl,
+			model.modelId,
+			requestBody,
+			apiToken,
+			parseOutput,
+			withModelLoading
+		);
+
+		isLoading = false;
+		// Reset values
+		computeTime = "";
+		error = "";
+		modelLoading = { isLoading: false, estimatedTime: 0 };
+		output = null;
+		outputJson = "";
+
+		if (res.status === "success") {
+			computeTime = res.computeTime;
+			output = res.output;
+			outputJson = res.outputJson;
+		} else if (res.status === "loading-model") {
+			modelLoading = {
+				isLoading: true,
+				estimatedTime: res.estimatedTime,
+			};
+			getOutput(true);
+		} else if (res.status === "error") {
+			error = res.error;
+		}
+	}
+
+	function isValidOutput(arg: any): arg is Output {
+		return (
+			arg &&
+			typeof arg === "object" &&
+			typeof arg["answer"] === "string" &&
+			Array.isArray(arg["coordinates"]) &&
+			Array.isArray(arg["cells"])
+		);
+	}
+
+	function parseOutput(body: unknown): Output | null {
+		return isValidOutput(body) ? body : null;
+	}
+
+	/*
+	 * Converts table from [[Header0, Header1, Header2], [Column0Val0, Column1Val0, Column2Val0], ...]
+	 * to {Header0: [ColumnVal0, ...], Header1: [Column1Val0, ...], Header2: [Column2Val0, ...]}
+	 */
+	function convertTableToData(table: string[][]): TableData {
+		return Object.fromEntries(
+			table[0].map((cell, x) => {
+				return [
+					cell,
+					table
+						.slice(1)
+						.flat()
+						.filter((_, i) => i % table[0].length === x)
+						.map((x) => String(x)), // some models can only handle strings (no numbers)
+				];
+			})
+		);
+	}
```

Comment:
I think we should refactor this. It's the same as Table QA no? You could put them 
in https://github.com/huggingface/huggingface_hub/blob/main/widgets/src/lib/InferenceWidget/shared/helpers.ts

Reply:
totally! should I put them [helpers.ts](https://github.com/huggingface/huggingface_hub/blob/main/widgets/src/lib/InferenceWidget/shared/helpers.ts) or [ViewUtils.ts](https://github.com/huggingface/huggingface_hub/blob/main/widgets/src/lib/InferenceWidget/shared/ViewUtils.ts)?

Reply:
Hmm ViewUtils seems like a better place